bolt-modules/boltlib/lib/puppet/datatypes/target.rb

```diff
@@ -1,22 +1,51 @@
 # frozen_string_literal: true
 
 Puppet::DataTypes.create_type('Target') do
-  interface <<-PUPPET
-    attributes => {
-      uri => String[1],
-      options => { type => Hash[String[1], Data], value => {} }
-    },
-    functions => {
-      name => Callable[[], String[1]],
-      host => Callable[[], Optional[String]],
-      password => Callable[[], Optional[String[1]]],
-      port => Callable[[], Optional[Integer]],
-      protocol => Callable[[], Optional[String[1]]],
-      user => Callable[[], Optional[String[1]]],
-    }
-  PUPPET
-
+  begin
+    inventory_version = Puppet.lookup(:bolt_inventory).version
+  rescue Puppet::Context::UndefinedBindingError
+    inventory_version = 1
+  end
   load_file('bolt/target')
 
-  implementation_class Bolt::Target
+  if inventory_version != 2
+    interface <<-PUPPET
+      attributes => {
+        uri => String[1],
+        options => { type => Hash[String[1], Data], value => {} }
+      },
+      functions => {
+        name => Callable[[], String[1]],
+        host => Callable[[], Optional[String]],
+        password => Callable[[], Optional[String[1]]],
+        port => Callable[[], Optional[Integer]],
+        protocol => Callable[[], Optional[String[1]]],
+        user => Callable[[], Optional[String[1]]],
+      }
+    PUPPET
+    implementation_class Bolt::Target
+  else
+    # Is the `kind` appropriate for these?
+    interface <<-PUPPET
+      attributes => {
+        uri => { type => Optional[String[1]], kind => given_or_derived },
+        name => { type => Optional[String[1]] , kind => given_or_derived },
+        target_alias => { type => Optional[Variant[String[1], Array[String[1]]]], kind => given_or_derived },
+        config => { type => Optional[Hash[String[1], Data]], kind => given_or_derived },
+        vars => { type => Optional[Hash[String[1], Data]], kind => given_or_derived  },
+        facts => { type => Optional[Hash[String[1], Data]], kind => given_or_derived  },
+        features => { type => Optional[Array[String[1]]], kind => given_or_derived },
+      },
+      functions => {
+        safe_name => Callable[[], String[1]],
+        host => Callable[[], Optional[String]],
+        password => Callable[[], Optional[String[1]]],
+        port => Callable[[], Optional[Integer]],
+        protocol => Callable[[], Optional[String[1]]],
+        user => Callable[[], Optional[String[1]]],
+      }
+    PUPPET
+
+    implementation_class Bolt::Target2
```

Comment:
In PE we need to be able to control Target::new() to disallow setting connection information, since that is expected to come from the Console inventory.

If this was `implementation_class inventory.target_class()`, we could override the Target class to our subclass without the need to modify this file further.

Reply:
/cc @McdonaldSeanp @adreyer

Reply:
Updated.